examples_tests/61.BlitFilterTest/main.cpp

```diff
@@ -0,0 +1,1561 @@
+		std::mt19937 prng;
+
+		uint8_t* bytePtr = reinterpret_cast<uint8_t*>(image->getBuffer()->getPointer());
+		for (uint64_t k = 0u; k < dims[2]; ++k)
+		{
+			for (uint64_t j = 0u; j < dims[1]; ++j)
+			{
+				for (uint64_t i = 0; i < dims[0]; ++i)
+				{
+					double decodedPixel[4] = { 0 };
+					for (uint32_t ch = 0u; ch < asset::getFormatChannelCount(format); ++ch)
+						decodedPixel[ch] = dist(prng);
+
+					const uint64_t pixelIndex = (k * dims[1] * dims[0]) + (j * dims[0]) + i;
+					asset::encodePixelsRuntime(format, bytePtr + pixelIndex * asset::getTexelOrBlockBytesize(format), decodedPixel);
+				}
+			}
+		}
+	}
+
+	return image;
+}
+
+static inline asset::IImageView<asset::ICPUImage>::E_TYPE getImageViewTypeFromImageType_CPU(const asset::IImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case asset::IImage::ET_1D:
+		return asset::ICPUImageView::ET_1D;
+	case asset::IImage::ET_2D:
+		return asset::ICPUImageView::ET_2D;
+	case asset::IImage::ET_3D:
+		return asset::ICPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<asset::IImageView<asset::ICPUImage>::E_TYPE>(0u);
+	}
+}
+
+static inline video::IGPUImageView::E_TYPE getImageViewTypeFromImageType_GPU(const video::IGPUImage::E_TYPE type)
+{
+	switch (type)
+	{
+	case video::IGPUImage::ET_1D:
+		return video::IGPUImageView::ET_1D;
+	case video::IGPUImage::ET_2D:
+		return video::IGPUImageView::ET_2D;
+	case video::IGPUImage::ET_3D:
+		return video::IGPUImageView::ET_3D;
+	default:
+		__debugbreak();
+		return static_cast<video::IGPUImageView::E_TYPE>(0u);
+	}
+}
+
+class CBlitFilter
+{
+private:
+	struct alignas(16) vec3_aligned
+	{
+		float x, y, z;
+	};
+
+	struct alignas(16) uvec3_aligned
+	{
+		uint32_t x, y, z;
+	};
+
+public:
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_DIM = 16u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_WORKGROUP_SIZE = 256u;
+	static constexpr uint32_t NBL_GLSL_DEFAULT_BIN_COUNT = 256u;
+
+	struct alpha_test_push_constants_t
+	{
+		float referenceAlpha;
+	};
+
+	struct blit_push_constants_t
+	{
+		uvec3_aligned inDim;
+		uvec3_aligned outDim;
+		vec3_aligned negativeSupport;
+		vec3_aligned positiveSupport;
+		uvec3_aligned windowDim;
+		uvec3_aligned phaseCount;
+		uint32_t windowsPerWG;
+		uint32_t axisCount;
+	};
+
+	struct normalization_push_constants_t
+	{
+		uvec3_aligned outDim;
+		uint32_t inPixelCount;
+		float referenceAlpha;
+	};
+
+	struct dispatch_info_t
+	{
+		uint32_t wgCount[3];
+	};
+
+	CBlitFilter(video::ILogicalDevice* logicalDevice, const uint32_t smemSize = 16*1024u) : device(logicalDevice), sharedMemorySize(smemSize)
+	{
+		sampler = nullptr;
+		{
+			video::IGPUSampler::SParams params = {};
+			params.TextureWrapU = asset::ISampler::ETC_CLAMP_TO_EDGE;
+			params.TextureWrapV = asset::ISampler::ETC_CLAMP_TO_EDGE;
+			params.TextureWrapW = asset::ISampler::ETC_CLAMP_TO_EDGE;
+			params.BorderColor = asset::ISampler::ETBC_FLOAT_OPAQUE_BLACK;
```

Comment:
`CBlitImageFilter` has these programmable
https://github.com/Devsh-Graphics-Programming/Nabla/blob/fd9be09ea1b0770f439554a1e2fd4035a7049e42/include/nbl/asset/filters/CBlitImageFilter.h#L47

Reply:
may I suggest having a local sampler cache (core::map from wrap modes and border color - should fit in a single uint16_t)?

Reply:
Is there a specific reason to not use `core::unordered_map`?

If we just use a simple flat array
```C++
core::smart_refctd_ptr<video::IGPUSampler> samplerCache[ETC_COUNT][ETC_COUNT][ETC_COUNT][ETBC_COUNT];
````
it would require around ~10kB of space --too much?

Reply:
fine, your call

Reply:
Hmm.. I think, since we use a combined image sampler, the sampler has to be baked into the descriptor set layout. We can add the border color and wrapping params as parameters to the existing cache for blit descriptor set layouts (which is currently only cached on `E_BLIT_TYPE`). Then we wouldn't need a separate cache just for samplers.
```C++
SBlitDescriptorSetLayoutCacheKey
{
    E_BLIT_TYPE blitType;
    video::IGPUSampler::E_TEXTURE_CLAMP wrapX;
    video::IGPUSampler::E_TEXTURE_CLAMP wrapY;
    video::IGPUSampler::E_TEXTURE_CLAMP wrapZ;
    video::IGPUSampler::E_TEXTURE_BORDER_COLOR borderColor;
};
```

Ofcourse then we would have to cache the blit pipeline layouts as well.

What do you think?

Reply:
> the sampler has to be baked into the descriptor set layout.

Not true, combined vs immutable are two separate concepts.

### Separate

`sampler2D` vs `texture2D` in the shader, and two SEPARATE descriptor bindings being used (one for image, one for sampler).

Combined Sampler puts the image and sampler into a single descriptor.

### Immutable

Whether the sampler you use is baked into the descriptor set, or written while you write the image descriptor.
We support choice between immutable and mutable, if you want mutable, you leave the sampler `nullptr` in the DS-Layout.

Reply:
Yeah you're right, I didn't know they were different concepts.

Reply:
Resolved.